enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md

```diff
@@ -0,0 +1,278 @@
+
+#### OpenShift components can't work with the CCM properly
+
+Since OpenShift imposes additional limitations compared to Kubernetes, some collisions are possible.
+So far there are no CCMs available in OpenShift, and there can be some problems with operability.
+
+Severity: medium-high
+Likelihood: low
+
+#### Cinder CSI driver is not available in OpenShift
+
+Integration with OpenStack Cinder (Persistent Volume Manager) was included in the in-tree cloud provider. But CCM doesn't contain this feature, so Cinder CSI driver must be installed in the system to work properly. Despite the fact that a cluster with some limitations can be installed without the driver, it will be impossible to work on it in production.
+
+Severity: medium-high
+Likelihood: low
+
+## Design Details
+
+### Test Plan
+
+Testing will consist in running of serial and parallel e2e-openstack jobs on the modified system with OpenStack CCM. If they pass successfully, we can consider the solution works.
+
+No changes to CI system are required, but additional tests should be run on a cloud with self-signed certificates.
+
+### Upgrade / Downgrade Strategy
+
+Upgrade from previous versions of OpenShift on OpenStack will look like:
+
+- On the initial stage of upgrading `cluster-version-operator` starts `cluster-cloud-controller-manager-operator`.
+
+- `cluster-cloud-controller-manager-operator` creates all required resources for CCM (Namespace, RBAC, Service Account).
+
+- `cluster-cloud-controller-manager-operator` creates a static pod manifest for CCM.
+
+- `kubelet` is restarted to start CCM in a static pod.
+
+- `kube-apiserver` and `kube-controller-manager` are restarted without the `--cloud-provider` option.
+
+- `kubelet` is restarted with `--cloud-provider external` option.
+
+Downgrade:
+
+- `kubelet` is restarted with `--cloud-provider openstack` option.
+
+- `kube-apiserver` and `kube-controller-manager` are restarted with the `--cloud-provider openstack` option.
+
+- `cluster-cloud-controller-manager-operator` destroys OpenStack CCM manifests and all related resources.
+
+- `kubelet` is restarted again.
+
+- `cluster-cloud-controller-manager-operator` stops working.
+
+### Version Skew Strategy
+
+See the upgrade/downgrade strategy.
+
+## Alternatives
+
+Only one alternative is to keep the deprecated in-tree cloud providers in OpenShift and support them with our own resources.
+
+## Infrastructure Needed
+
+Additional OpenStack cloud may be required to test how CCM works with self-signed certificates. Our current CI doesn't allow this.
+
+## Open questions
+
+1. Should we reuse the existing cloud provider config or generate a new one?
+CCM config is backward compatible with the in-tree cloud provider. It means we can reuse it.
+
+2. How to migrate PVs created by the in-tree cloud provider?
+[CSIMigration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) looks like the best option, especially if it is GA in 1.19.
+
```

Comment:
we need a section about bootstrapping. Probably we need a bootstrap instance of ccm.